pkg/helper/webconsole.go

```diff
@@ -0,0 +1,120 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"strings"
+	"time"
+)
+
+var logger = logs.GetLogger("openshift-webconsole")
+
+// ConsoleLinkText is the text of the consoleLink shown on the webconsole
+const ConsoleLinkText = "APIManager - 3scale"
+// ConsoleLinkNamePrefix is the prefix applied to Console link for system master
+const ConsoleLinkNamePrefix = "system-master-link"
+
+// ConsoleLinkSupported checks if a ConsoleLink CRD exists
+func ConsoleLinkSupported() error {
+	gvk := schema.GroupVersionKind{Group: "console.openshift.io", Version: "v1", Kind: "ConsoleLink"}
+	return kubernetes.CustomResourceDefinitionExists(gvk)
+}
+
+// ReconcileConsoleLink creates a ConsoleLink object if it doesn't already exist
+func ReconcileConsoleLink(ctx context.Context, c client.Client, req *reconcile.Request) {
+	route := getRoute(ctx, c, req)
+	if route == nil {
+		removeConsoleLink(ctx, c, req)
+	} else {
+		consoleLinkName := fmt.Sprintf("%s-%s", ConsoleLinkNamePrefix, req.Namespace)
+		consoleLink := &consolev1.ConsoleLink{}
+		err := c.Get(ctx, types.NamespacedName{Name: consoleLinkName}, consoleLink)
+		if err != nil && apierrors.IsNotFound(err) {
+			consoleLink = getConsoleLink(consoleLinkName, route, req)
+			if err := c.Create(ctx, consoleLink); err != nil {
+				logger.Error(err, "Console link is not created.")
+			} else {
+				logger.Info("Console link has been created:", consoleLinkName)
+			}
+		} else if err == nil && consoleLink != nil {
+			updateConsoleLink(ctx, route, consoleLink, c)
+		}
+	}
+}
+func getRoute(ctx context.Context, c client.Client, req *reconcile.Request) *routev1.Route {
+	// here the loop tries the best to collect a route, the reconcile event may comes
+	// before the route resources get created/updated
+	reader := read.New(c).WithNamespace(req.Namespace)
+	for i := 1; i < 30; i++ {
+		time.Sleep(time.Duration(100) * time.Microsecond)
+		deployedRoutes, err := reader.List(&routev1.RouteList{})
+		if err == nil {
+			for _, route := range deployedRoutes { //look for 3scale admin routes, capture route object
+				realr := route.(*routev1.Route)
+				if strings.Compare(realr.Spec.To.Name, "system-master") == 0 {
+					return realr
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func updateConsoleLink(ctx context.Context, route *routev1.Route, consoleLink *consolev1.ConsoleLink, c client.Client) {
+	url := "https://" + route.Spec.Host
+	linkTxt := ConsoleLinkText
+	if url != consoleLink.Spec.Href || linkTxt != consoleLink.Spec.Text {
+		consoleLink.Spec.Href = url
+		consoleLink.Spec.Text = linkTxt
+		if err := c.Update(ctx, consoleLink); err != nil {
+			logger.Error(err, "failed to reconcile Console Link", consoleLink)
+		}
+	}
+}
+
+func getConsoleLink(consoleLinkName string, route *routev1.Route, req *reconcile.Request) *consolev1.ConsoleLink {
```

Comment:
Could I get a captured picture of the ConsoleLink in the UI to see how it looks once created?

Reply:
![3sale_link](https://user-images.githubusercontent.com/42355256/94589176-f2434200-0252-11eb-960e-7474b45b2d14.png)

Reply:
Thanks for the capture :+1: . I see that it is an entry at project-level indeed.
I'm wondering whether we should be more specific in  displayed name of the link to make it clear that the url is the master portal url (as there are other urls: admin portal, developer portal, ...).

For example... Maybe:
"APIManager - 3scale Master Portal"
or
"APIManager - 3scale System Master Portal"
or
"APIManager - 3scale System Master"

What do you think @eguzki?

Reply:
`APIManager - 3scale Master Portal` seems to be a better option.

Reply:
changed